Add Fisheye calibration tool

DavidTorresOcana opened the pull request:
This PR introduces changes to be able to calibrate fisheye cameras and aim to Fix #146

This PR is related with ros/common_msgs#151 , ros-perception/vision_opencv#306 and ros-perception/image_common#146

Main changes it introduces are:
 * Upgrade current calibration tool to:
     - Allow user select between pinhole and fisheye model
     - Calibrate Monocular and Stereo with fisheye model

DavidTorresOcana committed:
Add Fisheye calibration tool

valgur commented on camera_calibration_fisheye/fiseye_cal_reference/mono/calibrate.py:
```diff
@@ -0,0 +1,95 @@
+import sys
```
There's a typo in the `fiseye_cal_reference` parent directory.

valgur commented:
Looks promising! I think I'll try it out on real data in a few days.
One high-level comment: could the new functionality be merged into the existing respective packages instead of creating new ones? Even better if it's possible to merge it into the existing nodes instead of adding new ones as well.

DavidTorresOcana commented:
Hello
With respect this calibration tool, I though to put it together but I do not know how to modify the application GUI. I hope somebody can help on this.
With respect calibration, when you try: Depending on the camera, results would vary. The general recommendation with fisheye is to **collect points on the corners of the image**.

DavidTorresOcana committed:
Merge remote-tracking branch 'upstream/melodic' into Branch_8d01c930e66bdeb58c4b138539a64ddea0b17e10

DavidTorresOcana committed:
Add calibration for fisheye cameras
Fix #146

DavidTorresOcana commented:
> Looks promising! I think I'll try it out on real data in a few days.
> One high-level comment: could the new functionality be merged into the existing respective packages instead of creating new ones? Even better if it's possible to merge it into the existing nodes instead of adding new ones as well.

I mange to find some time to look into this: Modifications done.

DavidTorresOcana committed:
Correct typo

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -238,7 +238,7 @@ def __init__(self, boards, flags=0, pattern=Patterns.Chessboard, name='',
         self.pattern = pattern
         self.br = cv_bridge.CvBridge()
-
+        self.distortion_model = "pinhole"
```
Making this a string isn't ideal, can you make an enum class?

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -517,13 +523,22 @@ def lrost(name, d, k, r, p, size):
 
     @staticmethod
-    def lryaml(name, d, k, r, p, size):
+    def lryaml(name, d, k, r, p, size,dist_model):
```
add a space between comma and `dist_model`

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -517,13 +523,22 @@ def lrost(name, d, k, r, p, size):
                     .replace("[", "").replace("]", "").replace("\n", "\n        ")
             ))
 
+        # Select dist model
```
You pass the distortion model in the function, you don't need these checks, just directly print.

DavidTorresOcana replied:
You need to find out the distortion model for formatting the message below

SteveMacenski replied:
Then you should create a function, right now you have the same block of code in 2 places

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -647,19 +661,28 @@ def cal_fromcorners(self, good):
+            self.distortion = dist_coeffs.flat[:8].reshape(-1, 1)
+        elif "fisheye" in self.distortion_model:
+            calibration_flags = cv2.fisheye.CALIB_FIX_SKEW + self.calib_flags # Add user flags
+            reproj_err, self.intrinsics, self.distortion, rvecs, tvecs = cv2.fisheye.calibrate(
```
You have some indentation problems here it appears

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -647,19 +661,28 @@ def cal_fromcorners(self, good):
+                     opts, ipts,
+                      self.size,
+                      intrinsics_in, None, flags = calibration_flags,
+                      criteria = (cv2.TERM_CRITERIA_EPS+cv2.TERM_CRITERIA_MAX_ITER, 30, 1e-6))
```
There's some hard coded numbers, should these be calib flags?

DavidTorresOcana replied:
There are pinhole and fisheye specific calibration flags. Any suggestion how to deal with this? Ideally we would need to offer the user the ability to choose flags

DavidTorresOcana replied:
Default criteria flags do provide good convergence, so not including them as seen in last commit.

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -674,14 +697,27 @@ def set_alpha(self, a):
+                    self.P[j,i] = ncm[j, i]
+            self.mapx, self.mapy = cv2.initUndistortRectifyMap(self.intrinsics, self.distortion, self.R, ncm, self.size, cv2.CV_32FC1)
+        elif "fisheye" in self.distortion_model:
+            # TODO: Implement re-computation of P given users balance value
```
This TODO seems pretty important, can you clarify why this isnt implemented or what the plan is?

DavidTorresOcana replied:
For some reason `fisheye.estimateNewCameraMatrixForUndistortRectify` does not provide a meaningful camera matrix. It may be just an Opencv bug
If anybody knows how to use this function please let me know. It only misses the capability for the user to select the balance.
Similar issue: https://stackoverflow.com/questions/34316306/opencv-fisheye-calibration-cuts-too-much-of-the-resulting-image

SteveMacenski replied:
The first answer in that thread has what seems to be a real work around, is that not usable here for some reason?

DavidTorresOcana replied:
A naive workaround was implemented by scaling the projection matrix focal lengths in eb285f3b131975ed7465cc75e86d5d59a6880081

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -704,7 +740,7 @@ def undistort_points(self, src):
     def as_message(self):
         """ Return the camera calibration as a CameraInfo message """
-        return self.lrmsg(self.distortion, self.intrinsics, self.R, self.P, self.size)
+        return self.lrmsg(self.distortion, self.intrinsics, self.R, self.P, self.size,self.distortion_model)
```
Space between comma

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -724,7 +760,7 @@ def ost(self):
 
     def yaml(self):
-        return self.lryaml(self.name, self.distortion, self.intrinsics, self.R, self.P, self.size)
+        return self.lryaml(self.name, self.distortion, self.intrinsics, self.R, self.P, self.size,self.distortion_model)
```
Space between comma

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -936,23 +972,39 @@ def cal_fromcorners(self, good):
+                                   flags = flags)
+        elif "fisheye" in self.distortion_model:
+            if LooseVersion(cv2.__version__).version[0] == 2:
+                print("You need OpenCV >3 to use fisheye camera model")
```
You should terminate after this log.

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -963,30 +1015,58 @@ def set_alpha(self, a):
 
-        return (self.lrmsg(self.l.distortion, self.l.intrinsics, self.l.R, self.l.P, self.size),
-                self.lrmsg(self.r.distortion, self.r.intrinsics, self.r.R, self.r.P, self.size))
+        return (self.lrmsg(self.l.distortion, self.l.intrinsics, self.l.R, self.l.P, self.size,self.l.distortion_model),
```
Space between comma both R and L

SteveMacenski reviewed (changes requested):
Just come cleanup. I also noticed you changed the permissions on a bunch of files, why is that necessary?

DavidTorresOcana committed:
Restore camera_calib files permisions

DavidTorresOcana committed:
Upgrades to calibrator tool for multi model calibration

DavidTorresOcana committed:
Solve fisheye balance selection TODO:
For some reason estimateNewCameraMatrixForUndistortRectify is not producing the expected result, hence a workaround was implemented

SteveMacenski commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -1014,7 +1096,7 @@ def ost(self):
 
     def yaml(self, suffix, info):
-        return self.lryaml(self.name + suffix, info.distortion, info.intrinsics, info.R, info.P, self.size)
+        return self.lryaml(self.name + suffix, info.distortion, info.intrinsics, info.R, info.P, self.size,self.camera_model)
```
Space after comma please

SteveMacenski reviewed (changes requested):
Last issue. Please also articulate how you tested this since there is no test coverage

DavidTorresOcana commented:
> Last issue. Please also articulate how you tested this since there is no test coverage

True: all these features are being tested by live calibration with a real camera.
I would encourage you or anybody else to very it too by calibrating a camera.

DavidTorresOcana committed:
Add fisheye calibration flags as user arguments

DavidTorresOcana committed:
Add undistortion of points for fisheye

DavidTorresOcana commented:
Commit bc4f2e3 introduces fisheye calibration flags as input arguments.
The proposed input arguments are:
``` 
Options:
  -h, --help            show this help message and exit
  -c CAMERA_NAME, --camera_name=CAMERA_NAME
                        name of the camera to appear in the calibration file

  Chessboard Options:
    You must specify one or more chessboards as pairs of --size and
    --square options.

    -p PATTERN, --pattern=PATTERN
                        calibration pattern to detect - 'chessboard',
                        'circles', 'acircles'
    -s SIZE, --size=SIZE
                        chessboard size as NxM, counting interior corners
                        (e.g. a standard chessboard is 7x7)
    -q SQUARE, --square=SQUARE
                        chessboard square size in meters

  ROS Communication Options:
    --approximate=APPROXIMATE
                        allow specified slop (in seconds) when pairing images
                        from unsynchronized stereo cameras
    --no-service-check  disable check for set_camera_info services at startup
    --queue-size=QUEUE_SIZE
                        image queue size (default 1, set to 0 for unlimited)

  Calibration Optimizer Options:
    --pinhole-fix-principal-point
                        for pinhole, fix the principal point at the image
                        center
    --pinhole-fix-aspect-ratio
                        for pinhole, enforce focal lengths (fx, fy) are equal
    --pinhole-zero-tangent-dist
                        for pinhole, set tangential distortion coefficients
                        (p1, p2) to zero
    --pinhole-k-coefficients=NUM_COEFFS
                        for pinhole, number of radial distortion coefficients
                        to use (up to 6, default 2)
    --fisheye-recompute-extrinsicsts
                        for fisheye, extrinsic will be recomputed after each
                        iteration of intrinsic optimization
    --fisheye-fix-skew  for fisheye, skew coefficient (alpha) is set to zero
                        and stay zero
    --fisheye-fix-principal-point
                        for fisheye,fix the principal point at the image
                        center
    --fisheye-k-coefficients=NUM_COEFFS
                        for fisheye, number of radial distortion coefficients
                        to use fixing to zero the rest (up to 4, default 4)
    --fisheye-check-conditions
                        for fisheye, the functions will check validity of
                        condition number
    --disable_calib_cb_fast_check
                        uses the CALIB_CB_FAST_CHECK flag for
                        findChessboardCorners
    --max-chessboard-speed=MAX_CHESSBOARD_SPEED
                        Do not use samples where the calibration pattern is
                        moving faster                      than this speed in
                        px/frame. Set to eg. 0.5 for rolling shutter cameras. 
```

SteveMacenski commented on camera_calibration/nodes/cameracalibrator.py:
```diff
@@ -73,24 +73,44 @@ def main():
     parser.add_option_group(group)
     group = OptionGroup(parser, "Calibration Optimizer Options")
-    group.add_option("--fix-principal-point",
+    group.add_option("--pinhole-fix-principal-point",
```
So this is something we can't do-- renaming the flags from this package. Those flags have been around for a very long time and it will break people's packages.

Update the help tag, but these names should be reverted. Adding the fisheye ones is good though

DavidTorresOcana replied:
Agree :+1 . Rolling back original flags in 6c50c6b574dc51cffa3ceb9d923fba6329c7f4da .

Proposed flags are now : 
```
Options:
  -h, --help            show this help message and exit
  -c CAMERA_NAME, --camera_name=CAMERA_NAME
                        name of the camera to appear in the calibration file

  Chessboard Options:
    You must specify one or more chessboards as pairs of --size and
    --square options.

    -p PATTERN, --pattern=PATTERN
                        calibration pattern to detect - 'chessboard',
                        'circles', 'acircles'
    -s SIZE, --size=SIZE
                        chessboard size as NxM, counting interior corners
                        (e.g. a standard chessboard is 7x7)
    -q SQUARE, --square=SQUARE
                        chessboard square size in meters

  ROS Communication Options:
    --approximate=APPROXIMATE
                        allow specified slop (in seconds) when pairing images
                        from unsynchronized stereo cameras
    --no-service-check  disable check for set_camera_info services at startup
    --queue-size=QUEUE_SIZE
                        image queue size (default 1, set to 0 for unlimited)

  Calibration Optimizer Options:
    --fix-principal-point
                        for pinhole, fix the principal point at the image
                        center
    --fix-aspect-ratio  for pinhole, enforce focal lengths (fx, fy) are equal
    --zero-tangent-dist
                        for pinhole, set tangential distortion coefficients
                        (p1, p2) to zero
    -k NUM_COEFFS, --k-coefficients=NUM_COEFFS
                        for pinhole, number of radial distortion coefficients
                        to use (up to 6, default 2)
    --fisheye-recompute-extrinsicsts
                        for fisheye, extrinsic will be recomputed after each
                        iteration of intrinsic optimization
    --fisheye-fix-skew  for fisheye, skew coefficient (alpha) is set to zero
                        and stay zero
    --fisheye-fix-principal-point
                        for fisheye,fix the principal point at the image
                        center
    --fisheye-k-coefficients=NUM_COEFFS
                        for fisheye, number of radial distortion coefficients
                        to use fixing to zero the rest (up to 4, default 4)
    --fisheye-check-conditions
                        for fisheye, the functions will check validity of
                        condition number
    --disable_calib_cb_fast_check
                        uses the CALIB_CB_FAST_CHECK flag for
                        findChessboardCorners
    --max-chessboard-speed=MAX_CHESSBOARD_SPEED
                        Do not use samples where the calibration pattern is
                        moving faster                      than this speed in
                        px/frame. Set to eg. 0.5 for rolling shutter cameras.

```

SteveMacenski reviewed (changes requested):
OK if tested on hardware I'm OK with that.

You broke backwards compatibility with the flag updates, we need another solution to that. I think just no qualifiers is fine since that's what OpenCV does anyhow and fisheye is specified

DavidTorresOcana committed:
cam_calib: rolling back flags
Rolling back changes to previous commit on camera calibrator flags to enable backwards compatibility

SteveMacenski reviewed (approved):
@JWhitleyAStuff take a look

JWhitleyWork commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -674,14 +699,26 @@ def set_alpha(self, a):
+            self.mapx, self.mapy = cv2.initUndistortRectifyMap(self.intrinsics, self.distortion, self.R, ncm, self.size, cv2.CV_32FC1)
+        elif self.camera_model == CAMERA_MODEL.FISHEYE:
+            # ncm = cv2.fisheye.estimateNewCameraMatrixForUndistortRectify(self.intrinsics, self.distortion, self.size, self.R, balance=a)
+            # self.P[:3,:3] = ncm[:3,:3]
```
Can you please remove the two lines above while leaving the note? The note seems enough to explain the method used without leaving the unused code.

DavidTorresOcana replied:
Agree

JWhitleyWork commented on camera_calibration/src/camera_calibration/calibrator.py:
```diff
@@ -647,19 +666,25 @@ def cal_fromcorners(self, good):
+                opts, ipts, self.size,
+                intrinsics_in, None, flags = self.fisheye_calib_flags)
+        else:
+            print("Something went wrong when selecting a model")
```
Instead of smattering this warning around everywhere, I would prefer that we guarantee that self.camera_model is valid when the parameter is set, otherwise exit with an error.

JWhitleyWork replied:
Also, I think the "more Pythonic" way is to ask for forgiveness rather than permission (AFRTP) instead of look before you leap (LBYL).

DavidTorresOcana replied:
It is guaranteed that self.camera_model is valid. That comment is not strictcly necessary for production and will be removed.

JWhitleyWork reviewed (changes requested):
See comments.

DavidTorresOcana committed:
cam_calib: Style formating
